File: test-repo/c/sample.c
/**
 * C sample file for AMP parser testing.
 * Tests: functions, structs, enums, typedefs, #include directives
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// Macro definitions
#define MAX_NAME_LENGTH 100
#define MAX_EMAIL_LENGTH 200
#define MAX_USERS 1000

// Enum definitions
typedef enum {
    STATUS_ACTIVE,
    STATUS_INACTIVE,
    STATUS_PENDING,
    STATUS_SUSPENDED
} Status;

typedef enum {
    ROLE_ADMIN,
    ROLE_USER,
    ROLE_GUEST
} UserRole;

// Struct definitions
typedef struct {
    long id;
    char name[MAX_NAME_LENGTH];
    char email[MAX_EMAIL_LENGTH];
    Status status;
    UserRole role;
    long created_at;
    long updated_at;
} User;

typedef struct {
    User* users[MAX_USERS];
    int count;
    long next_id;
} UserRepository;

typedef struct {
    const char* message;
    int code;
} Error;

// Function declarations
User* user_create(const char* name, const char* email);
void user_destroy(User* user);
bool user_validate(const User* user, Error* error);
bool user_is_active(const User* user);
void user_update_status(User* user, Status status);
void user_print(const User* user);

UserRepository* repository_create(void);
void repository_destroy(UserRepository* repo);
User* repository_save(UserRepository* repo, User* user);
User* repository_find_by_id(UserRepository* repo, long id);
bool repository_delete(UserRepository* repo, long id);
int repository_count(UserRepository* repo);
User** repository_find_active(UserRepository* repo, int* count);

bool validate_email(const char* email);
bool validate_name(const char* name);
long get_current_timestamp(void);
void log_info(const char* message);
void log_error(const char* message);

// User functions implementation
User* user_create(const char* name, const char* email) {
    User* user = (User*)malloc(sizeof(User));
    if (user == NULL) {
        return NULL;
    }
    
    user->id = 0;
    strncpy(user->name, name, MAX_NAME_LENGTH - 1);
    user->name[MAX_NAME_LENGTH - 1] = '\0';
    strncpy(user->email, email, MAX_EMAIL_LENGTH - 1);
    user->email[MAX_EMAIL_LENGTH - 1] = '\0';
    user->status = STATUS_ACTIVE;
    user->role = ROLE_USER;
    user->created_at = get_current_timestamp();
    user->updated_at = user->created_at;
    
    return user;
}

void user_destroy(User* user) {
    if (user != NULL) {
        free(user);
    }
}

bool user_validate(const User* user, Error* error) {
    if (user == NULL) {
        if (error != NULL) {
            error->message = "User is NULL";
            error->code = -1;
        }
        return false;
    }
    
    if (!validate_name(user->name)) {
        if (error != NULL) {
            error->message = "Invalid name";
            error->code = 1;
        }
        return false;
    }
    
    if (!validate_email(user->email)) {
        if (error != NULL) {
            error->message = "Invalid email";
            error->code = 2;
        }
        return false;
    }
    
    return true;
}

bool user_is_active(const User* user) {
    return user != NULL && user->status == STATUS_ACTIVE;
}

void user_update_status(User* user, Status status) {
    if (user != NULL) {
        user->status = status;
        user->updated_at = get_current_timestamp();
    }
}

void user_print(const User* user) {
    if (user == NULL) {
        printf("User is NULL\n");
        return;
    }
    
    printf("User{id=%ld, name='%s', email='%s', status=%d}\n",
           user->id, user->name, user->email, user->status);
}

// Repository functions implementation
UserRepository* repository_create(void) {
    UserRepository* repo = (UserRepository*)malloc(sizeof(UserRepository));
    if (repo == NULL) {
        return NULL;
    }
    
    repo->count = 0;
    repo->next_id = 1;
    
    for (int i = 0; i < MAX_USERS; i++) {
        repo->users[i] = NULL;
    }
    
    return repo;
}

void repository_destroy(UserRepository* repo) {
    if (repo == NULL) {
        return;
    }
    
    for (int i = 0; i < repo->count; i++) {
        if (repo->users[i] != NULL) {
            user_destroy(repo->users[i]);
        }
    }
    
    free(repo);
}

User* repository_save(UserRepository* repo, User* user) {
    if (repo == NULL || user == NULL) {
        return NULL;
    }
    
    Error error;
    if (!user_validate(user, &error)) {
        log_error(error.message);
        return NULL;
    }
    
    if (repo->count >= MAX_USERS) {
        log_error("Repository is full");
        return NULL;
    }
    
    if (user->id == 0) {
        user->id = repo->next_id++;
        user->created_at = get_current_timestamp();
    }
    
    user->updated_at = get_current_timestamp();
    repo->users[repo->count++] = user;
    
    char log_msg[256];
    snprintf(log_msg, sizeof(log_msg), "User saved: %s (ID: %ld)", user->name, user->id);
    log_info(log_msg);
    
    return user;
}

User* repository_find_by_id(UserRepository* repo, long id) {
    if (repo == NULL) {
        return NULL;
    }
    
    for (int i = 0; i < repo->count; i++) {
        if (repo->users[i] != NULL && repo->users[i]->id == id) {
            return repo->users[i];
        }
    }
    
    return NULL;
}

bool repository_delete(UserRepository* repo, long id) {
    if (repo == NULL) {
        return false;
    }
    
    for (int i = 0; i < repo->count; i++) {
        if (repo->users[i] != NULL && repo->users[i]->id == id) {
            user_destroy(repo->users[i]);
            
            // Shift remaining users
            for (int j = i; j < repo->count - 1; j++) {
                repo->users[j] = repo->users[j + 1];
            }
            
            repo->users[repo->count - 1] = NULL;
            repo->count--;
            
            return true;
        }
    }
    
    return false;
}

int repository_count(UserRepository* repo) {
    return repo != NULL ? repo->count : 0;
}

User** repository_find_active(UserRepository* repo, int* count) {
    if (repo == NULL || count == NULL) {
        return NULL;
    }
    
    User** active_users = (User**)malloc(sizeof(User*) * repo->count);
    if (active_users == NULL) {
        return NULL;
    }
    
    int active_count = 0;
    for (int i = 0; i < repo->count; i++) {
        if (user_is_active(repo->users[i])) {
            active_users[active_count++] = repo->users[i];
        }
    }
    
    *count = active_count;
    return active_users;
}

// Utility functions implementation
bool validate_email(const char* email) {
    if (email == NULL || strlen(email) == 0) {
        return false;
    }
    
    return strchr(email, '@') != NULL;
}

bool validate_name(const char* name) {
    if (name == NULL || strlen(name) < 2) {
        return false;
    }
    
    return true;
}

long get_current_timestamp(void) {
    return (long)time(NULL);
}

void log_info(const char* message) {
    printf("[INFO] %s\n", message);
}

void log_error(const char* message) {
    fprintf(stderr, "[ERROR] %s\n", message);
}

// Main function
int main(void) {
    UserRepository* repo = repository_create();
    if (repo == NULL) {
        log_error("Failed to create repository");
        return 1;
    }
    
    // Create users
    User* user1 = user_create("Alice", "alice@example.com");
    User* user2 = user_create("Bob", "bob@example.com");
    
    // Save users
    repository_save(repo, user1);
    repository_save(repo, user2);
    
    // Print all users
    printf("\nAll users:\n");
    for (int i = 0; i < repo->count; i++) {
        user_print(repo->users[i]);
    }
    
    // Find active users
    int active_count;
    User** active_users = repository_find_active(repo, &active_count);
    printf("\nActive users: %d\n", active_count);
    free(active_users);
    
    // Cleanup
    repository_destroy(repo);
    
    return 0;
}

File: test-repo/ruby/sample.rb
# Ruby sample file for AMP parser testing.
# Tests: classes, modules, methods, require/require_relative

require 'time'
require 'json'

# Module definitions
module AMP
  module TestRepo
    # Constants
    MAX_NAME_LENGTH = 100
    MAX_EMAIL_LENGTH = 200

    # Status module with constants
    module Status
      ACTIVE = 'active'
      INACTIVE = 'inactive'
      PENDING = 'pending'
      SUSPENDED = 'suspended'
    end

    # UserRole module
    module UserRole
      ADMIN = 'admin'
      USER = 'user'
      GUEST = 'guest'
    end

    # Validator module (mixin)
    module Validator
      def validate
        errors.empty?
      end

      def errors
        @errors ||= []
      end

      def add_error(message)
        errors << message
      end

      def clear_errors
        @errors = []
      end
    end

    # Logger module
    module Logger
      def log_info(message)
        puts "[INFO] #{Time.now.strftime('%Y-%m-%d %H:%M:%S')} - #{message}"
      end

      def log_error(message)
        warn "[ERROR] #{Time.now.strftime('%Y-%m-%d %H:%M:%S')} - #{message}"
      end

      def log_warning(message)
        puts "[WARN] #{Time.now.strftime('%Y-%m-%d %H:%M:%S')} - #{message}"
      end
    end

    # Base entity class
    class BaseEntity
      attr_accessor :id, :created_at, :updated_at

      def initialize
        @id = nil
        @created_at = Time.now
        @updated_at = Time.now
      end

      def valid?
        raise NotImplementedError, 'Subclasses must implement valid?'
      end

      def to_h
        {
          id: @id,
          created_at: @created_at.iso8601,
          updated_at: @updated_at.iso8601
        }
      end
    end

    # User class
    class User < BaseEntity
      include Validator

      attr_accessor :name, :email, :status, :role

      def initialize(name = nil, email = nil)
        super()
        @name = name
        @email = email
        @status = Status::ACTIVE
        @role = UserRole::USER
        @errors = []
      end

      def active?
        @status == Status::ACTIVE
      end

      def valid?
        clear_errors

        add_error('Name cannot be empty') if @name.nil? || @name.strip.empty?
        add_error('Invalid email format') if @email.nil? || !@email.include?('@')

        validate
      end

      def update_status(new_status)
        @status = new_status
        @updated_at = Time.now
      end

      def suspend(reason)
        @status = Status::SUSPENDED
        @updated_at = Time.now
        puts "User #{@name} suspended: #{reason}"
      end

      def promote_to_admin
        @role = UserRole::ADMIN
        @updated_at = Time.now
      end

      def to_h
        super.merge(
          name: @name,
          email: @email,
          status: @status,
          role: @role
        )
      end

      def to_json(*args)
        to_h.to_json(*args)
      end

      def to_s
        "User{id=#{@id}, name='#{@name}', email='#{@email}', status=#{@status}}"
      end
    end

    # Repository class
    class UserRepository
      include Logger

      def initialize
        @users = {}
        @next_id = 1
      end

      def find_by_id(id)
        @users[id]
      end

      def save(user)
        unless user.valid?
          raise ArgumentError, "Invalid user data: #{user.errors.join(', ')}"
        end

        if user.id.nil?
          user.id = @next_id
          @next_id += 1
          user.created_at = Time.now
        end

        user.updated_at = Time.now
        @users[user.id] = user

        log_info("User saved: #{user.name} (ID: #{user.id})")
        user
      end

      def delete(id)
        if @users.delete(id)
          log_info("User deleted: ID #{id}")
          true
        else
          false
        end
      end

      def find_all
        @users.values
      end

      def find_active_users
        @users.values.select(&:active?)
      end

      def find_by_role(role)
        @users.values.select { |user| user.role == role }
      end

      def count
        @users.size
      end

      def clear
        @users.clear
        @next_id = 1
      end
    end

    # Service class
    class UserService
      include Logger

      def initialize(repository)
        @repository = repository
      end

      def create_user(name, email)
        user = User.new(name, email)
        @repository.save(user)
      end

      def get_user_by_id(id)
        @repository.find_by_id(id)
      end

      def delete_user(id)
        @repository.delete(id)
      end

      def get_all_users
        @repository.find_all
      end

      def promote_to_admin(user_id)
        user = @repository.find_by_id(user_id)
        if user
          user.promote_to_admin
          @repository.save(user)
          log_info("User #{user.name} promoted to ADMIN")
        else
          log_error("User not found: ID #{user_id}")
        end
      end

      def suspend_user(user_id, reason)
        user = @repository.find_by_id(user_id)
        if user
          user.suspend(reason)
          @repository.save(user)
        else
          log_error("User not found: ID #{user_id}")
        end
      end

      def get_active_users
        @repository.find_active_users
      end
    end

    # Utility module
    module ValidationHelper
      module_function

      def valid_email?(email)
        !email.nil? && email.include?('@') && email.include?('.')
      end

      def valid_name?(name)
        !name.nil? && !name.strip.empty? && name.length >= 2
      end

      def sanitize_input(input)
        input.nil? ? '' : input.strip
      end
    end

    # Helper class with class methods
    class UserHelper
      class << self
        def format_user_display(user)
          "#{user.name} (#{user.email})"
        end

        def user_summary(users)
          {
            total: users.size,
            active: users.count(&:active?),
            inactive: users.count { |u| !u.active? }
          }
        end

        def export_to_json(users)
          users.map(&:to_h).to_json
        end
      end
    end

    # Custom error class
    class UserNotFoundError < StandardError
      attr_reader :user_id

      def initialize(user_id, message = 'User not found')
        @user_id = user_id
        super("#{message}: ID #{user_id}")
      end
    end
  end
end

# Main execution
if __FILE__ == $PROGRAM_NAME
  include AMP::TestRepo

  repository = UserRepository.new
  service = UserService.new(repository)

  # Create users
  user1 = service.create_user('Alice', 'alice@example.com')
  user2 = service.create_user('Bob', 'bob@example.com')

  puts "Created user: #{user1}"
  puts "Created user: #{user2}"

  # Promote user to admin
  service.promote_to_admin(user1.id)

  # List all users
  puts "\nAll users:"
  service.get_all_users.each do |user|
    puts "  #{user}"
  end

  # Get active users
  active_users = service.get_active_users
  puts "\nActive users: #{active_users.size}"

  # User summary
  summary = UserHelper.user_summary(service.get_all_users)
  puts "\nUser summary: #{summary}"
end
